### PR TITLE
[Feature][Connector-V2] Add influxDB connector source

### PR DESCRIPTION
**Pull request by 531651225**
## Purpose of this pull request
https://github.com/apache/incubator-seatunnel/issues/1946
Add InfluxDB Connector Source
close https://github.com/apache/incubator-seatunnel/pull/2595

## Check list

* [x] Code changed are covered with tests, or it does not need tests for reason:
* [x] If any new Jar binary package adding in your PR, please add License Notice according
  [New License Guide](https://github.com/apache/incubator-seatunnel/blob/dev/docs/en/contribution/new-license.md)
* [x] If necessary, please update the documentation to describe the new feature. https://github.com/apache/incubator-seatunnel/tree/dev/docs

### CODE REVIEW

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+        database = "test"
+        upper_bound = 100
+        lower_bound = 1
+        partition_num = 4
+        split_column = "value"
+        fields {
+            label = STRING
+            value = INT
+            rt = STRING
+            time = BIGINT
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf]: test all datatypes?

Comment (531651225) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf]: already fixed it

**Comment by hailin0** [comment]
can you add e2e-testcase(spark)?

**Comment by 531651225** [comment]
> can you add e2e-testcase(spark)?

there is a conflict of the okhttp jar  between the Spark jar  and  the influxdb connector.
At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document

**Comment by EricJoy2048** [comment]
> > can you add e2e-testcase(spark)?
> 
> there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document

@CalvinKirs  PTAL

**Comment by CalvinKirs** [comment]
> > > can you add e2e-testcase(spark)?
> > 
> > 
> > there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document
> 
> @CalvinKirs PTAL

We'd better be able to rename these jars @TyrantLucifer  WDYT?

**Comment by TyrantLucifer** [comment]
> > > > can you add e2e-testcase(spark)?
> > > 
> > > 
> > > there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document
> > 
> > 
> > @CalvinKirs PTAL
> 
> We'd better be able to rename these jars @TyrantLucifer WDYT?

Yes, aggre with @CalvinKirs , you can add `maven-shade` plugin in your pom to rename these jars. BTW, could you please use `httpclient`, not use `okhttp`, it's better to unify versions and dependencies for http dependencies. In subsequent iterations, we need to abstract this http-related functionality into public modules and replace them uniformly with the renamed httpclient. What do you think about? @531651225 @CalvinKirs

**Comment by 531651225** [comment]
> > > > > can you add e2e-testcase(spark)?
> > > > 
> > > > 
> > > > there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document
> > > 
> > > 
> > > @CalvinKirs PTAL
> > 
> > 
> > We'd better be able to rename these jars @TyrantLucifer WDYT?
> 
> Yes, aggre with @CalvinKirs , you can add `maven-shade` plugin in your pom to rename these jars. BTW, could you please use `httpclient`, not use `okhttp`, it's better to unify versions and dependencies for http dependencies. In subsequent iterations, we need to abstract this http-related functionality into public modules and replace them uniformly with the renamed httpclient. What do you think about? @531651225 @CalvinKirs

图片

It seems impossible to replace okhttp with httpclient. The okhttp jar is used internally by the influxdb-java.jar. 
Influxdb-java is the only way officially provided for accessing influxdb.

**Comment by TyrantLucifer** [comment]
> > > > > > can you add e2e-testcase(spark)?
> > > > > 
> > > > > 
> > > > > there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document
> > > > 
> > > > 
> > > > @CalvinKirs PTAL
> > > 
> > > 
> > > We'd better be able to rename these jars @TyrantLucifer WDYT?
> > 
> > 
> > Yes, aggre with @CalvinKirs , you can add `maven-shade` plugin in your pom to rename these jars. BTW, could you please use `httpclient`, not use `okhttp`, it's better to unify versions and dependencies for http dependencies. In subsequent iterations, we need to abstract this http-related functionality into public modules and replace them uniformly with the renamed httpclient. What do you think about? @531651225 @CalvinKirs
> 
> 图片
> 
> It seems impossible to replace okhttp with httpclient. The okhttp jar is used internally by the influxdb-java.jar. Influxdb-java is the only way officially provided for accessing influxdb.

This is really bad news. But don't worry, we also have the ultimate weapon, the `maven-shade` plugin. We can use this plugin to rename packages of okhttp to avoid jar conflict.

**Comment by 531651225** [comment]
> > > > > > > can you add e2e-testcase(spark)?
> > > > > > 
> > > > > > 
> > > > > > there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document
> > > > > 
> > > > > 
> > > > > @CalvinKirs PTAL
> > > > 
> > > > 
> > > > We'd better be able to rename these jars @TyrantLucifer WDYT?
> > > 
> > > 
> > > Yes, aggre with @CalvinKirs , you can add `maven-shade` plugin in your pom to rename these jars. BTW, could you please use `httpclient`, not use `okhttp`, it's better to unify versions and dependencies for http dependencies. In subsequent iterations, we need to abstract this http-related functionality into public modules and replace them uniformly with the renamed httpclient. What do you think about? @531651225 @CalvinKirs
> > 
> > 
> > 图片
> > It seems impossible to replace okhttp with httpclient. The okhttp jar is used internally by the influxdb-java.jar. Influxdb-java is the only way officially provided for accessing influxdb.
> 
> This is really bad news. But don't worry, we also have the ultimate weapon, the `maven-shade` plugin. We can use this plugin to rename packages of okhttp to avoid jar conflict.

Thanks for your advice. I'll try. By the way, is there any other case where the  shade plugin is used by other connectors

**Comment by TyrantLucifer** [comment]
> > > > > > > > can you add e2e-testcase(spark)?
> > > > > > > 
> > > > > > > 
> > > > > > > there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document
> > > > > > 
> > > > > > 
> > > > > > @CalvinKirs PTAL
> > > > > 
> > > > > 
> > > > > We'd better be able to rename these jars @TyrantLucifer WDYT?
> > > > 
> > > > 
> > > > Yes, aggre with @CalvinKirs , you can add `maven-shade` plugin in your pom to rename these jars. BTW, could you please use `httpclient`, not use `okhttp`, it's better to unify versions and dependencies for http dependencies. In subsequent iterations, we need to abstract this http-related functionality into public modules and replace them uniformly with the renamed httpclient. What do you think about? @531651225 @CalvinKirs
> > > 
> > > 
> > > 图片
> > > It seems impossible to replace okhttp with httpclient. The okhttp jar is used internally by the influxdb-java.jar. Influxdb-java is the only way officially provided for accessing influxdb.
> > 
> > 
> > This is really bad news. But don't worry, we also have the ultimate weapon, the `maven-shade` plugin. We can use this plugin to rename packages of okhttp to avoid jar conflict.
> 
> Thanks for your advice. I'll try. By the way, is there any other case where the shade plugin is used by other connectors

You can refer to https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html or see `seatunnel-config` module get the usage of shade plugin.

**Comment by 531651225** [comment]
> > > > > > > > > can you add e2e-testcase(spark)?
> > > > > > > > 
> > > > > > > > 
> > > > > > > > there is a conflict of the okhttp jar between the Spark jar and the influxdb connector. At present, this problem is solved by specifying the replacement of conflicting packages in the spark jars directory in the document
> > > > > > > 
> > > > > > > 
> > > > > > > @CalvinKirs PTAL
> > > > > > 
> > > > > > 
> > > > > > We'd better be able to rename these jars @TyrantLucifer WDYT?
> > > > > 
> > > > > 
> > > > > Yes, aggre with @CalvinKirs , you can add `maven-shade` plugin in your pom to rename these jars. BTW, could you please use `httpclient`, not use `okhttp`, it's better to unify versions and dependencies for http dependencies. In subsequent iterations, we need to abstract this http-related functionality into public modules and replace them uniformly with the renamed httpclient. What do you think about? @531651225 @CalvinKirs
> > > > 
> > > > 
> > > > 图片
> > > > It seems impossible to replace okhttp with httpclient. The okhttp jar is used internally by the influxdb-java.jar. Influxdb-java is the only way officially provided for accessing influxdb.
> > > 
> > > 
> > > This is really bad news. But don't worry, we also have the ultimate weapon, the `maven-shade` plugin. We can use this plugin to rename packages of okhttp to avoid jar conflict.
> > 
> > 
> > Thanks for your advice. I'll try. By the way, is there any other case where the shade plugin is used by other connectors
> 
> You can refer to https://maven.apache.org/plugins/maven-shade-plugin/examples/resource-transformers.html or see `seatunnel-config` module get the usage of shade plugin.

thks, The conflict has been resolved. 
have already added spark E2E, Please help  review   @hailin0 @CalvinKirs @EricJoy2048 @TyrantLucifer

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+        split_column = "f5"
+        fields {
+            label = STRING
+            f1 = STRING
+            f2 = DOUBLE
+            f3 = BIGINT
+            f4 = FLOAT
+            f5 = INT
+            f6 = SMALLINT
+            f7 = BOOLEAN
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf:47]: test all datatype?

reference
https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iceberg-flink-e2e/src/test/resources/iceberg/iceberg_source.conf#L30

File: seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-influxdb-spark-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf [hunk 3]
```diff
@@ -0,0 +1,180 @@
...
+    InfluxDB {
+        url = "http://influxdb-host:8086"
+        sql = "select label, f1, f2, f3, f4, f5, f6, f7 from test"
+        database = "test"
+        upper_bound = 100
+        lower_bound = 1
+        partition_num = 4
+        split_column = "f5"
+        fields {
+            label = STRING
```
Comment (hailin0) [inline seatunnel-e2e/seatunnel-spark-connector-v2-e2e/connector-influxdb-spark-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf:42]: as above

File: seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf (hunk 2 above)
Comment (531651225) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf:47]: Thank you for your suggestion, but at present, the only data types supported by influxdb are strings, floats, integers, or booleans. 

detail in https://docs.influxdata.com/influxdb/v1.8/concepts/glossary/#field-value
图片

Comment (531651225) [reply seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-influxdb-flink-e2e/src/test/resources/influxdb/influxdb_source_to_assert.conf:47]: > test all datatype?
> 
> reference https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-e2e/seatunnel-flink-connector-v2-e2e/connector-iceberg-flink-e2e/src/test/resources/iceberg/iceberg_source.conf#L30

> Thank you for your suggestion, but at present, the only data types supported by influxdb are strings, floats, integers, or booleans.
> 
> detail in https://docs.influxdata.com/influxdb/v1.8/concepts/glossary/#field-value 图片

**Comment by EricJoy2048** [comment]
@531651225  does this pr can review again? 
If it's ok,@hailin0 @TyrantLucifer  @Hisoka-X  please reviews this pr again.

**Comment by 531651225** [comment]
> @531651225 does this pr can review again? If it's ok,@hailin0 @TyrantLucifer @Hisoka-X please reviews this pr again.



> @531651225 does this pr can review again? If it's ok,@hailin0 @TyrantLucifer @Hisoka-X please reviews this pr again.

 Okay, help me review, thks

File: docs/en/connector-v2/source/InfluxDB.md [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+| database            | string  | yes      |               |
+| username    | string  | no       | -             |
+| password    | string  | no       | -             |
+| lower_bound | long    | no       | -             |
+| upper_bound | long    | no       | -             |
+| partition_num | int     | no       | -             |
+| split_column | string     | no       | -             |
+| epoch | string     | no       | n             |
+| connect_timeout_ms | long     | no       | 15000         |
+| query_timeout_sec | int     | no       | 3             |
```
Comment (hailin0) [inline docs/en/connector-v2/source/InfluxDB.md]: check style

**Review by hailin0 — COMMENTED** [review]
LGTM

File: docs/en/connector-v2/source/InfluxDB.md (hunk 4 above)
Comment (531651225) [reply docs/en/connector-v2/source/InfluxDB.md]: > check style

already fixed

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java [hunk 5]
```diff
@@ -0,0 +1,65 @@
...
+        SeaTunnelDataType<?>[] seaTunnelDataTypes = typeInfo.getFieldTypes();
+
+        for (int i = 0; i <= seaTunnelDataTypes.length - 1; i++) {
+            Object seatunnelField;
+            int columnIndex = indexList.get(i);
+            SeaTunnelDataType<?> seaTunnelDataType = seaTunnelDataTypes[i];
+            if (null == values.get(columnIndex)) {
+                seatunnelField = null;
+            }
+            else if (BasicType.BOOLEAN_TYPE.equals(seaTunnelDataType)) {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java]: use `seaTunnelDataType.getSqlType()` better.

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+    }
+
+    public InfluxDBSourceSplitEnumerator(SourceSplitEnumerator.Context<InfluxDBSourceSplit> context, InfluxDBSourceState sourceState, InfluxDBConfig config) {
+        this(context, config);
+        this.assignedSplit = sourceState.getAssignedSplit();
+    }
+
+    @Override
+    public void open() {
+        this.assignedSplit = new HashSet<>();
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java]: `open()` will be called again after `public InfluxDBSourceSplitEnumerator(SourceSplitEnumerator.Context<InfluxDBSourceSplit> context, InfluxDBSourceState sourceState, InfluxDBConfig config)`.

Therefore, we cannot reset split in open method.

Comment (hailin0) [reply seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java]: reference
https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/source/IoTDBSourceSplitEnumerator.java#L83

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java (hunk 5 above)
Comment (531651225) [reply seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java]: already fixed

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java (hunk 6 above)
Comment (531651225) [reply seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java]: already fixed

Comment (531651225) [reply seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java]: > `open()` will be called again after `public InfluxDBSourceSplitEnumerator(SourceSplitEnumerator.Context<InfluxDBSourceSplit> context, InfluxDBSourceState sourceState, InfluxDBConfig config)`.
> 
> Therefore, we cannot reset split in open method.

already fixed @hailin0 @EricJoy2048

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxdbSourceReader.java [hunk 7]
```diff
@@ -0,0 +1,150 @@
...
+        if (Boundedness.BOUNDED.equals(context.getBoundedness())) {
+            // signal to the source that we have reached the end of the data.
+            log.info("Closed the bounded fake source");
+            context.signalNoMoreElement();
+        }
+    }
+
+    @Override
+    public List<InfluxDBSourceSplit> snapshotState(long checkpointId) throws Exception {
+        return null;
```
Comment (ashulin) [inline seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxdbSourceReader.java]: Please add split snapshot.
you can see https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/source/IoTDBSourceReader.java

Comment (531651225) [reply seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxdbSourceReader.java]: > Please add split snapshot. you can see https://github.com/apache/incubator-seatunnel/blob/dev/seatunnel-connectors-v2/connector-iotdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/iotdb/source/IoTDBSourceReader.java

already fixed

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+import org.apache.seatunnel.api.table.type.SqlType;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class InfluxDBRowConverter {
+
+    public static SeaTunnelRow convert(List<Object> values, SeaTunnelRowType typeInfo, List<Integer> indexList) {
+
+        List<Object> fields = new ArrayList<>();
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java]: Gave a init size better.

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java [hunk 9]
```diff
@@ -0,0 +1,215 @@
...
+        }
+    }
+
+    @Override
+    public void run() {
+        Set<Integer> readers = context.registeredReaders();
+        if (shouldEnumerate) {
+            Set<InfluxDBSourceSplit> newSplits = getInfluxDBSplit();
+
+            synchronized (stateLock) {
```
Comment (EricJoy2048) [inline seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/source/InfluxDBSourceSplitEnumerator.java:67]: @ashulin @hailin0   PTAL

File: seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java (hunk 8 above)
Comment (531651225) [reply seatunnel-connectors-v2/connector-influxdb/src/main/java/org/apache/seatunnel/connectors/seatunnel/influxdb/converter/InfluxDBRowConverter.java]: > Gave a init size better.

have already fixed

**Review by ashulin — APPROVED** [review]
LGTM

**Review by EricJoy2048 — APPROVED** [review]
+1

**Pull request merged**